### PR TITLE
feat: use CAREN to set default cert SANS

### PR DESCRIPTION
**Pull request by faiq**
This replaces https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/20

This adds the logic to add necessary default Extra Sans to the `KubeadmControlPlane` for Docker Clusters via a mutation hook.

### CODE REVIEW

**Comment by faiq** [comment]
Copying a comment over, for posterity and to continue discussion 


@supershal 
>  > WDYT of creating new separate patch dockerExtraAPIServerCertSANsPatch for Docker MetaVariable. Ensure that both lifecycle patch extraAPIServerCertSANsPatch and docker patchdockerExtraAPIServerCertSANsPatch preserves existing list of certSANs. i.e. something like

> A bit more details: I think it would be cleaner to have provider specific logic in the provider's handler. i.e. pkg/handlers/docker/mutation and generic handlers can be free of any logic that is based on provider. Having a docker mutator that always adds default SANs could be cleaner IMHO

@faiq 
> I disagree. I think adding another handler that touches the same exact fields on objects can be more prone to errors and harder for a developer to discover what gets applied when

File: pkg/handlers/generic/mutation/extraapiservercertsans/inject.go [hunk 1]
```diff
@@ -64,11 +79,13 @@ func (h *extraAPIServerCertSANsPatchHandler) Mutate(
...
 		return err
 	}
-	if !found {
-		log.V(5).Info("Extra API server cert SANs variable not defined")
+	if !found && len(defaultAPICertSANs) == 0 {
+		log.V(5).Info("No Extra API server cert SANs needed to be added")
 		return nil
 	}
 
+	extraSans := append(extraAPIServerCertSANsVar, defaultAPICertSANs...)
```
Comment (dkoshkin) [inline pkg/handlers/generic/mutation/extraapiservercertsans/inject.go:87]: nit, the "extraSans" being a combination of "extraAPIServerCertSANsVar" and "defaultAPICertSANs" is a little confusing, wdyt of just `apiServerCertSANsVar`?

```suggestion
	apiServerCertSANsVar := append(extraAPIServerCertSANsVar, defaultAPICertSANs...)
```

File: pkg/handlers/generic/mutation/extraapiservercertsans/inject.go [hunk 2]
```diff
@@ -83,15 +100,27 @@ func (h *extraAPIServerCertSANsPatchHandler) Mutate(
...
 			}
-			obj.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer.CertSANs = extraAPIServerCertSANsVar
-
+			obj.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer.CertSANs = extraSans
 			return nil
 		},
 	)
 }
+
+func getDefaultAPIServerSANs(cluster *capiv1.Cluster) []string {
```
Comment (dkoshkin) [inline pkg/handlers/generic/mutation/extraapiservercertsans/inject.go:115]: We now also have it in the Nutanix CC, please handle that one too

https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/blob/15224874e594e057403b7d521aabd40a44dc5053/charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml#L124-L128

File: pkg/handlers/generic/mutation/extraapiservercertsans/inject.go [hunk 3]
```diff
@@ -83,15 +100,27 @@ func (h *extraAPIServerCertSANsPatchHandler) Mutate(
...
+	provider, ok := cluster.Labels[capiv1.ProviderNameLabel]
+	if !ok {
+		return []string{}
+	}
+	switch provider {
+	case "docker":
```
Comment (dkoshkin) [inline pkg/handlers/generic/mutation/extraapiservercertsans/inject.go:121]:
```suggestion
	case v1alpha1.DockerProvider:
```

File: pkg/handlers/generic/mutation/extraapiservercertsans/inject.go (hunk 2 above)
Comment (faiq) [reply pkg/handlers/generic/mutation/extraapiservercertsans/inject.go]: I'll stack to this implementation.

File: pkg/handlers/generic/mutation/extraapiservercertsans/inject.go (hunk 1 above)
Comment (faiq) [reply pkg/handlers/generic/mutation/extraapiservercertsans/inject.go]: i like it!

**Comment by supershal** [comment]
>I disagree. I think adding another handler that touches the same exact fields on objects can be more prone to errors and harder for a developer to discover what gets applied when

Please check an alternative: https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/compare/faiq/docker-sans...shalin/docker-sans . It was able to add default cert sans for docker even if `clusterConfig.docker`  variable was not set for a docker cluster.  
Also changed existing `extraAPIServerCertSANsPatchHandler` to ensure it preserves any cert SANs set by other handlers.
We should discuss if multiple handlers should mutate same fields on object or not

**Pull request closed without merging**